File: leetcode2/MinimumPossibleIntegerAfteratMostKAdjacentSwapsOnDigits/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <string>
#include <algorithm>

using namespace std;

// Time Limit Exceed
// DFS
// O(NlgN) O(N)
class Solution {
 public:
  string minInteger(string s, int k) {
    // base
    if (k <= 0)
      return s;
    int n = s.size();
    if (k >= n*(n+1)/2) {
      sort(s.begin(), s.end());
      return s;
    }
    // recursion
    for (int i = 0; i < 10; i++) {
      int idx = s.find(to_string(i));
      if (idx >= 0 && idx <= k) {
        return s[idx] + minInteger(s.substr(0, idx) +
                                   s.substr(idx+1), k-idx);
      }
    }
    return s;
  }
};

File: leetcode2/MinimumPossibleIntegerAfteratMostKAdjacentSwapsOnDigits/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 316ms 85.64% 484.7MB 6.08%
class Solution:
	def minInteger(self, s: str, k: int) -> str:
		# base
		if k <= 0:
			return s
		n = len(s)
		if k >= n*(n+1)/2:
			return "".join(sorted(list(s)))
		for i in range(10):
			idx = s.find(str(i))
			if 0 <= idx <= k:
				return str(s[idx]) + self.minInteger(s[0:idx] + s[idx+1:], k-idx)
				